mono/metadata/icall.c

```diff
@@ -250,6 +253,7 @@ ves_icall_System_Array_SetValueImpl (MonoArray *arr, MonoObject *value, guint32
 
 #define INVALID_CAST G_STMT_START{ \
 		mono_get_runtime_callbacks ()->set_cast_details (vc, ec); \
+		mono_error_cleanup (&error);				\
```

Comment:
We should raise the underlying exception and not hide it under an InvalidCastException